### ISSUE TITLE
Cannot access to attribute if it setted through magic method

### ISSUE THREAD

**Issue by microcoder** [microsoft/pyright#113]
**Describe the bug**
Pyright cannot access to attribute of class if initialization was through the magic method `__setattr__`

**To Reproduce**
Create a class and initialize attributes through the method `__setattr__`

```python
class Demo():

    def __init__(self, **kwargs):
        super().__setattr__('_data', kwargs or {})
        print('After initialization:', self._data)
```
Then I get the Pyright error of access to the attribute:

```
_data : Unknown
Cannot access member '_data' on type 'Demo' pyright
```

**VS Code extension or command-line**
VS Code extension - v.1.0.24

**Comment by erictraut** [comment]
I don't consider this a bug. Setting an attribute by passing a string value means that you are manipulating the object in a dynamic manner. A static type checker will not know about object manipulations of this type. If you want to use static type checking, you should avoid this pattern.

Is there any reason the following doesn't meet your needs?
```
class Demo():
    def __init__(self, **kwargs):
        self._data = kwargs
```

**Comment by microcoder** [comment]
> I don't consider this a bug.

But this is valid code on Python. 

> Is there any reason the following doesn't meet your needs?

OK, look at the following code:

```python
class ClassA():

    def __setattr__(self, name, value):
        print('__setattr__ inside ClassA')

        # Some a handler...
        if name == '_data':
            value['name'] = 'A'

        super().__setattr__(name, value)


class ClassB(ClassA):

    def __init__(self, **kwargs):
        if True:  # some check
            super().__setattr__('_data', kwargs)
            self._some_method(self._data)  # Cannot access member '_data' on type 'ClassB' pyright
        else:
            self._data = kwargs

    def _some_method(self, name: str):
        pass


b = ClassB(name='B', param1=1, param2=2)
print(b._data)  # Cannot access member '_data' on type 'ClassB' pyright
```

**Comment by erictraut** [comment]
Anything piece of text can be considered "valid code on Python" in that a Python interpreter will attempt to execute it. A static type checker needs to make certain assumptions about what is "correct" and what isn't.

Pyright assumes that you are not dynamically modifying objects by inserting extra members into them. If it didn't make this assumption, it would need to relax its type-checking rules to such an extent that it would be of little value.

I'm curious why you see the need to use `super().__setattr__('_data', kwargs)` rather than `self._data = kwargs`. Is there a good reason?

You are still seeing the error above because pyright is statically determining that the "else" clause is never executed, so it's ignoring it. You would need to use a dynamic expression to defeat pyright's static analysis here. For example, define a constant `NOT_TYPE_CHECKING = True` and change your if statement to `if NOT_TYPE_CHECKING:`.

**Comment by erictraut** [comment]
If there's a good reason to support dynamic addition and removal of members from a particular class, you can define a custom `__getattr__` magic method for the class. This tells pyright to suspend its normal assumptions for that class. In your example above, `ClassA` could define a `__getattr__` method.

**Comment by microcoder** [comment]
Ah, sorry! The code which I inserted, it few incorrect for demonstration. I fixed the code:

```python
class ClassA():

    def __setattr__(self, name, value):
        print('__setattr__ inside ClassA')

        # Some a handler...
        if name == '_data':
            value['name'] = 'A'

        super().__setattr__(name, value)


class ClassB(ClassA):

    def __init__(self, **kwargs):
        if True:  # some check
            super().__setattr__('_data', kwargs)  # Initialization attribute only at ClassA
            # self._data = kwargs                 # Initialization attribute at ClassA and ClassB
            self._some_method()
        else:
            self._data = kwargs

    def __setattr__(self, name, value):
        print('__setattr__ inside ClassB')

        super().__setattr__(name, value)

    def _some_method(self):
        # Here is some handler of _data:
        self._data   # Cannot access member '_data' on type 'ClassB' pyright


b = ClassB(name='B', param1=1, param2=2)
print(b._data)  # Cannot access member '_data' on type 'ClassB' pyright
```
Initialization attribute only at ClassA, result:

```
__setattr__ inside ClassA
{'name': 'A', 'param1': 1, 'param2': 2}
```

Since the `kwargs` type is known as `Dict`, can Pyright assume that `type(self._data) == type (kwargs)`?

**Comment by microcoder** [comment]
> you can define a custom `__getattr__` magic method for the class. This tells pyright to suspend its normal assumptions for that class.

Hmm.. Ok. Thank you!

**Comment by erictraut** [comment]
If you assign an expression to an instance member and there is no explicit type declaration on that instance member, pyright will infer the type from the expression on the right-hand side of the assignment statement. In your example, kwargs has no declared type, so the inferred type of `kwargs` (and therefore of `self._data`) is `Dict[str, Any]`. If all of the kwargs dictionary entries have the same type, you can add a type annotation to the `**kwargs` parameter. For example, if you say `**kwargs: float`, pyright will know that the effective type is `Dict[str, float]`.

**Comment by ljw1004** [comment]
Hi hit another case of this. [playground](https://pyright-play.net/?code=FAYwNghgzlAEDCAuYtWwCYFMBmtsHt8AKKTMbASlgFoA%2BWAOXwDtNk0PYAHaKYYLLgBGEAE4kylGvSat2aHjH6kALhBUrx8ADQAiEaN3aDFfiFgBeBEVMgAdAWK27Bm8CA)
```

class C:
    def foo(self) -> None:
        pass

def bar(self) -> None:
    pass

setattr(C,"bar",bar)

c = C()
c.foo()
c.bar()  # error
```
